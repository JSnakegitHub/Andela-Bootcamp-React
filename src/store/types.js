export const FETCH_QUESTION = "FETCH_QUESTION";
export const FETCH_QUESTIONS = "FETCH_QUESTIONS";
export const POST_QUESTION = "POST_QUESTION";
export const LOGIN_USER = "LOGIN_USER";
export const SIGN_UP_USER = "SIGN_UP_USER";
