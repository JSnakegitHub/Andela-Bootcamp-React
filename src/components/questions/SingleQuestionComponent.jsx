import { connect } from "react-redux";
import React, { Component } from "react";
import { NavLink } from "react-router-dom";
import { fetchQuestions, fetchQuestion } from "../../store/actions/questions";
export class SingleQuestionComponent extends Component {
  constructor(props) {
    super(props);
  }
  componentWillMount() {
    this.props.fetchQuestions();
  }

  render() {
    function truncate(str, no_words) {
      return str
        .split(" ")
        .splice(0, no_words)
        .join(" ");
    }
    const fetchedQuestions = this.props.questions.map(question => (
      <div className="card mt-3 question-body" key={question.question_id}>
        <div className="card-body">
          <div className="single-question">
            <div className="row">
              <div className="col-md-1 text-center spacing">
                1 <br />
                Votes
              </div>
              <div className="col-md-1 text-center spacing">
                2 <br />
                Answers
              </div>
              <div className="col-md-10">
                {question.question_title}
                <div className="row">
                  <div className="col-md-6">
                    {truncate(question.question_body, 5)}...{" "}
                    <NavLink
                      className="navbar-brand"
                      to={`/articles?token=${question.question_id}`}
                      onClick={this.handleSingleQuestion}
                    >
                      read more
                    </NavLink>
                  </div>
                  <div className="col-md-6 text-right">
                    Asked by {question.question_author}
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    ));
    return <div>{fetchedQuestions}</div>;
  }
}
export const mapStateToProps = state => ({
  questions: state.questions.questions
});

export const mapDispatchToProps = dispatch => ({
  fetchQuestions: () => dispatch(fetchQuestions()),
  fetchQuestion: data => dispatch(fetchQuestion(data))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(SingleQuestionComponent);
