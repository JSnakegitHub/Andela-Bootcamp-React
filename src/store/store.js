import { createStore, applyMiddleware, compose } from "redux";
import thunk from "redux-thunk";
import rootReducer from "./reducers/RootReducer";
export default createStore(rootReducer, applyMiddleware(thunk));
