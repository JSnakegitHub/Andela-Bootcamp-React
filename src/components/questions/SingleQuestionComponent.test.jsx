import React from "react";
import { shallow, mount } from "enzyme";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import thunk from "redux-thunk";
import questionsReducer from "../../store/reducers/questions";
import { FETCH_QUESTIONS, POST_QUESTION } from "../../store/types";
import SingleQuestionComponent, {
  SingleQuestionComponent as SingleQuestionComponentTest
} from "./SingleQuestionComponent";

// const initialState = { questions: { questions: {} } };
const mockStore = configureStore([thunk])({ questions: { questions: {} } });
describe("", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(
      <MemoryRouter>
        <Provider store={mockStore}>
          <SingleQuestionComponent
            fetchQuestions={jest.fn}
            fetchQuestion={jest.fn}
            questions={[
              {
                question_ask_date: "12/12/2018",
                question_author: "hh",
                question_body: "The explanation is here",
                question_id: 24,
                question_title: "My Own Question"
              }
            ]}
          />
        </Provider>
      </MemoryRouter>
    );
  });

  it("should call handleSubmit method ", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should return the initial state", () => {
    expect(questionsReducer(undefined, {})).toEqual({
      question: {},
      questions: []
    });
  });

  it("should handle FETCH_QUESTIONS", () => {
    const startAction = {
      type: FETCH_QUESTIONS
    };

    expect(questionsReducer(null, startAction)).toEqual({});
  });

  it("should handle POST_QUESTION", () => {
    const payload = {
      question_title: "ghghgh",
      question_body: "pshdfhdh",
      question_author: "nyeko1"
    };

    const startAction = {
      type: POST_QUESTION,
      payload
    };

    expect(questionsReducer(null, startAction)).toEqual({ question: payload });
  });
});
