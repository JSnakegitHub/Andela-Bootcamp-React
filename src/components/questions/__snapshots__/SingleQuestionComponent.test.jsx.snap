// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should call handleSubmit method  1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MemoryRouter>
    <Provider
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
    >
      <Connect(SingleQuestionComponent)
        fetchQuestion={[Function]}
        fetchQuestions={[Function]}
        questions={
          Array [
            Object {
              "question_ask_date": "12/12/2018",
              "question_author": "hh",
              "question_body": "The explanation is here",
              "question_id": 24,
              "question_title": "My Own Question",
            },
          ]
        }
      />
    </Provider>
  </MemoryRouter>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <Provider
        store={
          Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
        }
      >
        <Connect(SingleQuestionComponent)
          fetchQuestion={[Function]}
          fetchQuestions={[Function]}
          questions={
            Array [
              Object {
                "question_ask_date": "12/12/2018",
                "question_author": "hh",
                "question_body": "The explanation is here",
                "question_id": 24,
                "question_title": "My Own Question",
              },
            ]
          }
        />
      </Provider>,
      "history": Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "2o4465",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "2o4465",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <Connect(SingleQuestionComponent)
          fetchQuestion={[Function]}
          fetchQuestions={[Function]}
          questions={
            Array [
              Object {
                "question_ask_date": "12/12/2018",
                "question_author": "hh",
                "question_body": "The explanation is here",
                "question_id": 24,
                "question_title": "My Own Question",
              },
            ]
          }
        />,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "fetchQuestion": [Function],
          "fetchQuestions": [Function],
          "questions": Array [
            Object {
              "question_ask_date": "12/12/2018",
              "question_author": "hh",
              "question_body": "The explanation is here",
              "question_id": 24,
              "question_title": "My Own Question",
            },
          ],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <Provider
          store={
            Object {
              "clearActions": [Function],
              "dispatch": [Function],
              "getActions": [Function],
              "getState": [Function],
              "replaceReducer": [Function],
              "subscribe": [Function],
            }
          }
        >
          <Connect(SingleQuestionComponent)
            fetchQuestion={[Function]}
            fetchQuestions={[Function]}
            questions={
              Array [
                Object {
                  "question_ask_date": "12/12/2018",
                  "question_author": "hh",
                  "question_body": "The explanation is here",
                  "question_id": 24,
                  "question_title": "My Own Question",
                },
              ]
            }
          />
        </Provider>,
        "history": Object {
          "action": "POP",
          "block": [Function],
          "canGo": [Function],
          "createHref": [Function],
          "entries": Array [
            Object {
              "hash": "",
              "key": "2o4465",
              "pathname": "/",
              "search": "",
              "state": undefined,
            },
          ],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "index": 0,
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "key": "2o4465",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <Connect(SingleQuestionComponent)
            fetchQuestion={[Function]}
            fetchQuestions={[Function]}
            questions={
              Array [
                Object {
                  "question_ask_date": "12/12/2018",
                  "question_author": "hh",
                  "question_body": "The explanation is here",
                  "question_id": 24,
                  "question_title": "My Own Question",
                },
              ]
            }
          />,
          "store": Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          },
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "fetchQuestion": [Function],
            "fetchQuestions": [Function],
            "questions": Array [
              Object {
                "question_ask_date": "12/12/2018",
                "question_author": "hh",
                "question_body": "The explanation is here",
                "question_id": 24,
                "question_title": "My Own Question",
              },
            ],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
